examples/tracing/lbr.py

```diff
@@ -0,0 +1,267 @@
+#!/usr/bin/python
+#
+# lbr.py
+#
+# Trace conditional branches executed by syscalls using the Last Branch Record
+# Buffer (LBR)
+#
+# REQUIRES:
+#   Linux 5.16+ (bpf_get_branch_snapshot support)
+#
+# Copyright (c) 2023 Bytedance Inc.
+#
+# Author(s):
+#   Lorenzo Carrozzo <lorenzocarrozzo@bytedance.com>
+
+from __future__ import absolute_import, print_function, unicode_literals
+from bcc import BPF, utils
+from bcc import PerfType, PerfHWConfig, PerfEventSampleFormat, Perf
+import argparse
+from sys import exit
+from pathlib import Path
+from subprocess import Popen, PIPE
+
+# Number of LBR entries and output tags
+lbr_cnt = 20  # Should be 32 but limited by bpf stack size
```

Comment:
I think we can create a percpu map with one element and the element size can be big enough to support 32 entries. This will make example more useful to people wanting to get lbr record in their bpf programs.

Reply:
Ive tried doing something similar in the past like so.
`struct lbr_entries {
    struct perf_branch_entry entries[ENTRY_CNT];
};

BPF_PERCPU_ARRAY(pc_arr, struct lbr_entries, 1);

// Function to use with tracepoint
int disp_snapshot_tp(struct params *p) {
    int zero = 0;
    if (pc_arr.lookup(&zero) == NULL) {
        return 1;
    }
    struct lbr_entries *l_entries = pc_arr.lookup(&zero);
    struct perf_branch_entry *entries = l_entries->entries;

    int total_entries = bpf_get_branch_snapshot(entries, sizeof(*entries), 0);
    total_entries /= sizeof(struct perf_branch_entry);

    if (true T_R_COND P_COND) {

        bpf_trace_printk("NUM_ENTRIES%d", total_entries);

        for (int i = 0; i < ENTRY_CNT; i++) {
            bpf_trace_printk("ENTRY%pS --> %pS", entries[i].from,
                             entries[i].to);
        }
    }

    return 0;
}`

But I get the following error which Im not sure how to interpret.
![Screenshot 2024-01-05 at 10 44 42](https://github.com/iovisor/bcc/assets/152497231/423a42dc-f5d8-4f1f-9b3f-1173680d681b)

Reply:
In the kernel self test gets around the limited stack size issue by initialising the array of entries as a global variable.
https://github.com/torvalds/linux/blob/master/tools/testing/selftests/bpf/progs/get_branch_snapshot.c

Is there a way of replicating this method in BCC as the per cpu array does not seem to work?

Reply:
The following code works for me:

```
diff --git a/examples/tracing/lbr.py b/examples/tracing/lbr.py
index 3d77051b..633d3bef 100644
--- a/examples/tracing/lbr.py
+++ b/examples/tracing/lbr.py
@@ -22,7 +22,7 @@ from pathlib import Path
 from subprocess import Popen, PIPE
 
 # Number of LBR entries and output tags
-lbr_cnt = 20  # Should be 32 but limited by bpf stack size
+lbr_cnt = 32
 num_entries_tag = 'lbr_total_entries:'
 entry_tag = 'lbr_entry:'
 
@@ -41,11 +41,22 @@ struct params {
         long ret;
 };
 
+struct perf_branch_entry_buf {
+    struct perf_branch_entry entries[ENTRY_CNT];
+};
+
+BPF_PERCPU_ARRAY(branch_entry, struct perf_branch_entry_buf, 1);
+
 // Function to use with tracepoint
 int disp_snapshot_tp(struct params *p) {
-    struct perf_branch_entry entries[ENTRY_CNT] = {};
+    unsigned buf_size = sizeof(struct perf_branch_entry_buf), idx = 0;
+    struct perf_branch_entry_buf *buf;
 
-    int total_entries = bpf_get_branch_snapshot(entries, sizeof(entries), 0);
+    buf = branch_entry.lookup(&idx);
+    if (!buf)
+        return 0;
+
+    int total_entries = bpf_get_branch_snapshot(buf, buf_size, 0);
     total_entries /= sizeof(struct perf_branch_entry);
 
     if (true T_R_COND P_COND) {
@@ -53,8 +64,8 @@ int disp_snapshot_tp(struct params *p) {
         bpf_trace_printk("NUM_ENTRIES%d", total_entries);
 
         for (int i = 0; i < ENTRY_CNT; i++) {
-            bpf_trace_printk("ENTRY%pS --> %pS", entries[i].from,
-                             entries[i].to);
+            bpf_trace_printk("ENTRY%pS --> %pS", buf->entries[i].from,
+                             buf->entries[i].to);
         }
     }
 
@@ -63,9 +74,14 @@ int disp_snapshot_tp(struct params *p) {
 
 // Function to use with kretprobe
 int disp_snapshot_krp(struct pt_regs *p) {
-    struct perf_branch_entry entries[ENTRY_CNT] = {};
+    unsigned buf_size = sizeof(struct perf_branch_entry_buf), idx = 0;
+    struct perf_branch_entry_buf *buf;
+
+    buf = branch_entry.lookup(&idx);
+    if (!buf)
+        return 0;
 
-    int total_entries = bpf_get_branch_snapshot(entries, sizeof(entries), 0);
+    int total_entries = bpf_get_branch_snapshot(buf, buf_size, 0);
     total_entries /= sizeof(struct perf_branch_entry);
 
     if (true K_R_COND P_COND) {
@@ -73,8 +89,8 @@ int disp_snapshot_krp(struct pt_regs *p) {
         bpf_trace_printk("NUM_ENTRIES%d", total_entries);
 
         for (int i = 0; i < ENTRY_CNT; i++) {
-            bpf_trace_printk("ENTRY%pS --> %pS", entries[i].from,
-                             entries[i].to);
+            bpf_trace_printk("ENTRY%pS --> %pS", buf->entries[i].from,
+                             buf->entries[i].to);
         }
     }
```

To trace nanosleep syscall,
```
...
i  | T/F  | Addresses                            |                                                                                                            
-------------------------------------------------|                                                                                                            
0  | From | trace_call_bpf+0x76/0x240            |                                                                                                            
   |---------------------------------------------|                                                                                                            
   | To   | trace_call_bpf+0xb6/0x240            |                                                                                                            
-------------------------------------------------|                                                                                                            
1  | From | perf_trace_buf_alloc+0x38/0xb0       |                                                                                                            
   |---------------------------------------------|                                                                                                            
   | To   | perf_trace_buf_alloc+0x53/0xb0       |                                                                                                            
-------------------------------------------------|                                                                                                            
2  | From | perf_syscall_exit+0x3b/0x160         |                                                                                                            
   |---------------------------------------------|                                                                                                            
   | To   | perf_syscall_exit+0x64/0x160         | 
...
-------------------------------------------------|                             
13 | From | finish_task_switch+0x1d1/0x2d0       |                             
   |---------------------------------------------|                             
   | To   | finish_task_switch+0x7d/0x2d0        |                             
-------------------------------------------------|                             
14 | From | perf_pmu_sched_task+0xfd/0x130       |                             
   |---------------------------------------------|                             
   | To   | perf_pmu_sched_task+0x102/0x130      |                             
-------------------------------------------------|                             
15 | From | 0x0                                  |                             
   |---------------------------------------------|                             
   | To   | 0x0                                  |                             
-------------------------------------------------|                             
16 | From | 0x0                                  |                             
   |---------------------------------------------|                             
   | To   | 0x0                                  |                             
-------------------------------------------------|     
...
-------------------------------------------------|                             
30 | From | 0x0                                  |                             
   |---------------------------------------------|                             
   | To   | 0x0                                  |                             
-------------------------------------------------|                             
31 | From | 0x0                                  |                             
   |---------------------------------------------|                             
   | To   | 0x0                                  |                             
-------------------------------------------------|  
```

If the above does not work, then your kernel probably too old.

Reply:
While the changes I had tried weren't fully correct. It seems like the error was generated as I was checking that the that the first item in the percpu array was not null rather than checking that the pointer was not null.

I have added the correct changes now to the patch.